packages/ruleset/src/functions/use-date-based-format.js

```diff
@@ -0,0 +1,401 @@
+/**
+ * Copyright 2024 IBM Corporation.
+ * SPDX-License-Identifier: Apache2.0
+ */
+
+const {
+  getExamplesForSchema,
+  isDateSchema,
+  isDateTimeSchema,
+  isIntegerSchema,
+  isObject,
+  isStringSchema,
+  validateNestedSchemas,
+} = require('@ibm-cloud/openapi-ruleset-utilities');
+
+const {
+  LoggerFactory,
+  isDateBasedName,
+  isDateBasedValue,
+  isParamContentSchema,
+  isParamSchema,
+  isRequestBodySchema,
+  isResponseSchema,
+  isSchemaProperty,
+} = require('../utils');
+
+let ruleId;
+let logger;
+
+/**
+ * The implementation for this rule makes assumptions that are dependent on the
+ * presence of the following other rules:
+ *
+ * - oas3-valid-media-example
+ * - oas3-valid-schema-example
+ *
+ * These rules verify that the correct, specific format (date vs date-time) is
+ * used for schemas based on their example value. So, we aren't as specific
+ * with that check in this rule - we recommend either "date" or "date-time".
+ */
+
+module.exports = function (schema, _opts, context) {
+  if (!logger) {
+    ruleId = context.rule.name;
+    logger = LoggerFactory.getInstance().getLogger(ruleId);
+  }
+
+  return checkForDateBasedFormat(
+    schema,
+    context.path,
+    context.documentInventory.resolved
+  );
+};
+
+/**
+ * This function implements a rule that enforces date-based schemas use either
+ * the "date" or "date-time" format, so that they're accurately documented as
+ * date-based logical types. We use a heuristic based on either the name of a
+ * schema (derived from the property name, if the schema is a property schema)
+ * or the example value provided for a given schema or schema property.
+ *
+ * The logic here recursively checks all schemas for the presence of unmarked
+ * date-based schemas. As it traverses the schemas, it compiles a list of
+ * potentially-relevant example values. This way, if an object schema defines
+ * its own example, which includes a value for a nested property that should
+ * be identified by the rule, we can track down the value once we reach the
+ * schema for said property. The logic will also gather any relevant parameter
+ * or media type examples that may be defined outside of the schema path.
+ *
+ * @param {object} s the schema to check
+ * @param {array} p the array of path segments indicating the "location" of the schema within the API definition
+ * @param {object} apidef the resolved API definition
+ * @returns an array containing the violations found or [] if no violations
+ */
+function checkForDateBasedFormat(s, p, apidef) {
+  // Map connecting a list of examples for a schema to its logical path.
+  const examples = {};
+
+  // Check for any examples outside of the schema path - they may be in
+  // request bodies, response bodies, or parameters. Store these separately.
+  const indirectExamples = checkForIndirectExamples(p, apidef);
```

Comment:
Personally, I think it's better to always prioritize examples closer to the schema definition because then the example cited in the rule output will be easier to find when looking at the schema it pertains to.

With that being said, I think `indirectExamples` should probably simply be prepended to `examples` for primary schemas.

Reply:
> With that being said, I think indirectExamples should probably simply be prepended to examples for primary schemas.

This is how I implemented it initially when updating the logic. However, I adjusted it for two reasons.

1. The OpenAPI specification states that 

> if referencing a schema which contains an example, the example(s) value SHALL override the example provided by the schema.

so I thought this logic would keep things closer to the specification (not that it really matters for this heuristic).

2. Separating it out made it much easier to log when an example value was found on an indirect example vs a primary schema example.

> Personally, I think it's better to always prioritize examples closer to the schema definition because then the example cited in the rule output will be easier to find when looking at the schema it pertains to.

I do agree that it makes it easier for the user. Let me know what you think about the above context.